packages/babel-plugin-transform-typescript/src/index.js

```diff
@@ -0,0 +1,266 @@
+import syntaxTypeScript from "babel-plugin-syntax-typescript";
+
+import transpileEnum from "./enum";
+
+function visitPattern({ node }) {
+  if (node.typeAnnotation) node.typeAnnotation = null;
+  if (node.type === "Identifier" && node.optional) node.optional = null;
+  // 'access' and 'readonly' are only for parameter properties, so constructor visitor will handle them.
+}
+
+function isInType(path) {
+  switch (path.parent.type) {
+    case "TSTypeReference":
+    case "TSQualifiedName":
+    case "TSExpressionWithTypeArguments":
+    case "TSTypeQuery":
+      return true;
+    default:
+      return false;
+  }
+}
+
+function isImportTypeOnly(binding, sourceFileHasJsx) {
+  for (const path of binding.referencePaths) {
+    if (!isInType(path)) {
+      return false;
+    }
+  }
+
+  if (binding.identifier.name != "React") {
+    return true;
+  }
+
+  return !sourceFileHasJsx;
+}
+
+interface State {
+  sourceFileHasJsx: boolean,
+}
+
+export default function({ types: t }) {
+  return {
+    inherits: syntaxTypeScript,
+    visitor: {
+      //"Pattern" alias doesn't include Identifier or RestElement.
+      Pattern: visitPattern,
+      Identifier: visitPattern,
+      RestElement: visitPattern,
+
+      Program(path, state: State) {
+        state.sourceFileHasJsx = false;
+        path.traverse({
```

Comment:
We can avoid a full file traversal if we only do this when we encounter a `import React`.